Data/ByteString/Lazy/ReadInt.hs

```diff
@@ -0,0 +1,380 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilyDependencies #-}
+
+-- We need @AllowAmbiguousTypes@ in order to be able to use @TypeApplications@
+-- to disambiguate the desired instance of class methods whose instance cannot
+-- be inferred from the caller's context.  We would otherwise have to use
+-- proxy arguments.  In particular, this is used for the Word64-valued upper
+-- and lower-bounds of various integral types ('wmax64', 'imax64' and 'imin64').
```

Comment:
If that's all, it might be nicer to just move these constants into `_readInt` and `_readWord`. But it's not a big deal.

Reply:
There'd multiple flavours of them, one for each output size.  I choice to make the functions polymorphic, to write them just once, but then inlined, where'd they'd be automatically specialised to the appropriate type, and presumably computed just once.  I'll check the core code to see whether that's actually what happens.

Reply:
Yes, and each flavour is used exactly once, in the corresponding specialization of `_readInt` or `_readWord`. And were they local variables within the body of these functions, `imax64` and friends can just have the (unambiguous) type `Word64`. I would expect the numeric values in each specialization to be evaluated at compile-time either way.

Reply:
Sorry, that doesn't work, because the various `readInt<N>` functions all call a single polymorphic `_readInt`, which uses polymorphic `imax64` and `imin64`, but it is specialised to all the sizes.  I don't think this warrants further attention unless the values to be coputed at compile time in the specialisations...

Reply:
I think we don't understand each other. I mean putting `imin64` in the `where` block attached to `_readInt`, and likewise for the other two. What do you mean?

(Your claim that "that doesn't work" was the last straw, getting me to finally successfully build bytestring from source to verify that my proposed alternative _does_ work. Thank you? I also found `GHC.Prim.leWord# x_aZQT 32768##` as the scrutinee of a case-expression with with `GHC.Maybe.Nothing @ (Int16, ByteString)` as a branch body in the resulting Core from both versions (up to a different variable name), which I assume confirms GHC had no trouble evaluating the `imin64` or `imin64 @ Int16` in `_readInt @ Int16` in `readInt16` at compile-time.)

Reply:
You're of course right, sorry wasn't paying attention.  Fixed along with all other review comments so far in the `fixup` commit (to be squashed post review).

The new `readNatural` is no longer quadratic.